node-repository/src/main/java/com/yahoo/vespa/hosted/provision/restapi/v2/NodePatcher.java

```diff
@@ -30,19 +31,26 @@
  */
 public class NodePatcher {
 
-    public static final String HARDWARE_FAILURE_DESCRIPTION = "hardwareFailureDescription";
+    private static final String HARDWARE_FAILURE_DESCRIPTION = "hardwareFailureDescription";
+    private static final String WANT_TO_RETIRE = "wantToRetire";
+
     private final NodeFlavors nodeFlavors;
     private final Inspector inspector;
     private final NodeRepository nodeRepository;
 
     private Node node;
+    private List<Node> children;
```

Comment:
This and above field should be `final` to avoid this class having internal state.

Reply:
These problems seem to exist because there's an unclear distinction of the arguments to this class and the result created in `apply`.

How about building a new list in `apply` (e.g. `modifiedNodes`) and return that? Something like:

```
    public List<Node> apply() {
        List<Node> modifiedNodes = new ArrayList<>();
        inspector.traverse((String name, Inspector value) -> {
            try {
                modifiedNodes.add(applyField(name, value));
            } catch (IllegalArgumentException e) {
                throw new IllegalArgumentException("Could not set field '" + name + "'", e);
            }
            modifiedNodes.addAll(applyFieldRecursive(children, name, value));
        });
        return Collections.unmodifiableList(modifiedNodes);
    }
```

Reply:
This wont work because for each traverse, I want to use the `node` and `children` that resulted from the last traverse. 

If there is a patch that updates two fields, the code above would produce `modifiedNodes` that contains `node` with only field 1 applied, and `node` with only field 2 applied. Same for the `children`, so `modifiedNodes.size() == 2* (1 + children.size())`.

Reply:
Right, so the `node` class member is changed for each field update? That's terrible. 😞 

Feel free to merge, but I think this should be rewritten at some point.